### PR TITLE
CP-10774: Add an 'install-supp-pack' plugin

### PR DESCRIPTION
**Pull request by jonludlam**
Signed-off-by: Jon Ludlam <jonathan.ludlam@citrix.com>

### CODE REVIEW

File: scripts/plugins/install-supp-pack [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    def __init__(self, *args):
+        PluginError.__init__(self, *args)
+
+def install(session, args):
+    if not args.has_key("vdi"):
+        raise ArgumentError("Key 'vdi' missing from arguments")
+    vdi=args['vdi']
+    process = subprocess.Popen([
+        "/opt/xensource/debug/with-vdi %s /opt/xensource/bin/xe-install-supplemental-pack" % vdi],
+        shell = True,
```
Comment (simonjbeaumont) [inline scripts/plugins/install-supp-pack]: According to the Python docs, this is a security hole. See the big red warning section here: https://docs.python.org/2/library/subprocess.html#frequently-used-arguments

File: scripts/plugins/install-supp-pack [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python
+
+import subprocess
+
+# Simple XenAPI plugin
+import XenAPIPlugin, time
```
Comment (simonjbeaumont) [inline scripts/plugins/install-supp-pack]: PEP-8 would say these should be on separate lines...

File: scripts/plugins/install-supp-pack [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+class PluginError(Exception):
+    """Base Exception class for all plugin errors."""
+    def __init__(self, *args):
+        Exception.__init__(self, *args)
+
+class ArgumentError(PluginError):
+    """Raised when the arguments to the plugin are invalid"""
+
```
Comment (simonjbeaumont) [inline scripts/plugins/install-supp-pack]: Blank line can be removed?

File: scripts/plugins/install-supp-pack [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    def __init__(self, *args):
+        PluginError.__init__(self, *args)
+
+class SubprocessFailure(PluginError):
+    """Raised when there was a failure attaching the VDI"""
+
```
Comment (simonjbeaumont) [inline scripts/plugins/install-supp-pack]: ditto.

File: scripts/plugins/install-supp-pack (hunk 1 above)
Comment (thomassa) [inline scripts/plugins/install-supp-pack]: So we should use `shell = False`, or we should ensure that `vdi` is harmless (check that it matches the format of a uuid), or preferably both (for defence in depth).

**Comment by simonjbeaumont** [comment]
Looks OK, now!

**Pull request merged**